### PR TITLE
enable portname plugin by default, turn some of sjy96525's changes into .po files

### PR DESCRIPTION
**Pull request by DavidHaywood**
note, we probably need a better way to do this so that we don't have to duplicate or create files for every clone
my original suggestion is that we require a list of sets that the replacements apply to above each list of replacements and process 'drivername.po' as well as 'parent.po' to look for replacements.  Also having an 'import from' type logic might help keep things tidier too.

if we're going to push this as the official way to do button naming for the more obvious cases then we really need it enabled by default, which is why I made that change, otherwise it's never going to gain any momentum.

of course this whole thing is up for review.

### CODE REVIEW

**Comment by cracyc** [comment]
Here's what I replied to you on the original commit:

> or similar (as mash suggests) before the list of replacement names would help so that it knows exactly which sets to apply the list of replacements that follows to without having to have a file for each. Not listing any sets would generate an error.

The problem I see here is that there are child sets that share a port definition with each other but not with the parent. How about a comment in the first line of a file like that contains a redirect to another file that will be used like "#redirect spclords.po" in spclordsa.po otherwise use the parent if there is no po file at all.

File: src/emu/emuopts.cpp [hunk 1]
```diff
@@ -209,7 +209,7 @@ const options_entry emu_options::s_option_entries[] =
 	{ OPTION_AUTOBOOT_SCRIPT ";script",                  nullptr,     OPTION_STRING,     "lua script to execute after machine boot" },
 	{ OPTION_CONSOLE,                                    "0",         OPTION_BOOLEAN,    "enable emulator LUA console" },
 	{ OPTION_PLUGINS,                                    "1",         OPTION_BOOLEAN,    "enable LUA plugin support" },
-	{ OPTION_PLUGIN,                                     nullptr,     OPTION_STRING,     "list of plugins to enable" },
+	{ OPTION_PLUGIN,                                     "portname",  OPTION_STRING,     "list of plugins to enable" },
```
Comment (cracyc) [inline src/emu/emuopts.cpp]: Don't do this.  Just change portname/plugin.json "start" to "true".

Comment (DavidHaywood) [reply src/emu/emuopts.cpp]: ahh, didn't know there was a better way :-)

**Comment by DavidHaywood** [comment]
for the children that share names but parents don't I'm not sure I see the problem

if each list of replacements has to have the sets to which it applies above it then you'd list the parent above one set of replacements, and the clones above another, depending on exactly what each applies to.

that way it can open the parent.po file, or drivername.po file and still only load the list of replacements for the sets in question.

let's say there was a namcos12.po or tekken3.po and it contained the following

tekken3:
msgid "P1 Button 1"
msgstr "P1 Left Punch"

msgid "P1 Button 2"
msgstr "P1 Right Punch"

msgid "P1 Button 3"
msgstr "P1 Button 3 (Unused)"

msgid "P1 Button 4"
msgstr "P1 Left Kick"

msgid "P1 Button 5"
msgstr "P1 Right Kick"

msgid "P1 Button 6"
msgstr "P1 Button 6 (Unused)"

msgid "P2 Button 1"
msgstr "P2 Left Punch"

msgid "P2 Button 2"
msgstr "P2 Right Punch"

msgid "P2 Button 3"
msgstr "P2 Button 3 (Unused)"

msgid "P2 Button 4"
msgstr "P2 Left Kick"

msgid "P2 Button 5"
msgstr "P2 Right Kick"

msgid "P2 Button 6"
msgstr "P2 Button 6 (Unused)"


tekken3a,tekken3b:
msgid "P1 Button 1"
msgstr "P1 Left Kick"

msgid "P1 Button 2"
msgstr "P1 Right Kick"

msgid "P1 Button 3"
msgstr "P1 Button 3 (Unused)"

msgid "P1 Button 4"
msgstr "P1 Left Punch"

msgid "P1 Button 5"
msgstr "P1 Right Punch"

msgid "P1 Button 6"
msgstr "P1 Button 6 (Unused)"

msgid "P2 Button 1"
msgstr "P2 Left Kick"

msgid "P2 Button 2"
msgstr "P2 Right Kick"

msgid "P2 Button 3"
msgstr "P2 Button 3 (Unused)"

msgid "P2 Button 4"
msgstr "P2 Left Punch"

msgid "P2 Button 5"
msgstr "P2 Right Punch"

msgid "P2 Button 6"
msgstr "P2 Button 6 (Unused)"

(this isn't a real example, but let's suppose the clones have kick/punch swapped)

the list for tekken3 would be processed if the actual set launched was tekken3, the lists for tekken3a/b would be processed if the set launched was tekken3a or b.  It's not a huge amount of extra data to process.

redirects / imports would be handy if only a few buttons change, eg. for tekkentag it could redirect to tekken3, because actually the buttonnames are the same but tekkentag needs the 'tag' name adding.

**Comment by cracyc** [comment]
Okay, how about this.  If the first line of a file is "#import <filename>" it'll process that file first then the rest of the current file.  It'll process a file starting with the first entries that have no set tag which are common for every set in the file.  If the set is a parent, after that there can a series of "#set <set1>,<set2>,.." (prepened with a # so po file editors will ignore them as comments) for all the sets which use the following entries.  That should make it possible to handle lots of overlap without being too complex.  Softlist entries will be same except they will look for setname_softname.po first.

**Comment by DavidHaywood** [comment]
Well as long as the system means you don't need a separate file for every single clone then that's the main issue avoided, separate files for every clone is much less manageable, especially once you take the softlists into account.

I think the best thing we can do is see how it goes.

**Comment by happppp** [comment]
I don't think this is a good solution to the portnames. The MAME source is supposed to be documentation. If we end up putting all button names in files like these, it will become very unclear. Browsing the source code, you'd get completely lost trying to figure out which input port pin is which button. I'd rather see a combination of tags(eg. "IN0") and masks, same how artwork layouts reference a button.

**Comment by stilett0** [comment]
Hello, Master of the Obvious here again.

So there's some previous art in this area of providing "names" for "buttons" in external files, so here's some data points.

The frontend community knows this functionality of naming, and counting, buttons and controls in an external file known as "controls.dat" - but the files in question are really an XML file, and an INI file for even older frontends, and that was previously hosted at controls.arcadecontrols.com before they underwent some website issues.
Archived here: https://web.archive.org/web/20150220112407id_/http://controls.arcadecontrols.com/

You can find the last known version of the files here:
http://ledblinky.net/Download.htm
or here:
controls.ini.0.141.1.zip -> http://www.mediafire.com/file/b5ublb5ccw4x3ch/controls.ini.0.141.1.zip
controls.xml.0.141.1.zip -> http://www.mediafire.com/file/xwnb1pk6e5b7fnn/controls.xml.0.141.1.zip
FEDEV.0.141.1.zip -> http://www.mediafire.com/file/52ibs0i2m4b5yey/FEDEV.0.141.1.zip

There's another Windows-friendly editor for this format here:
http://headsoft.com.au/index.php?category=controlsdat

Related stuff towards the display of control panels, buttons, etc. in MAME with this utility here:
http://headsoft.com.au/index.php?category=cpwizard
Remarkably, a use of MAME's Lua plugin spec outside of the MAME codebase.

---

Discussion of Controls.xml/Controls.ini continue here, if sporadically:
http://forum.arcadecontrols.com/index.php/board,45.0.html

Interesting/relevant threads that I recommend anyone implementing some sort of conversion/import/backwards support check out:
Is controls.dat dead?: http://forum.arcadecontrols.com/index.php/topic,149453.0.html
controls.json: http://forum.arcadecontrols.com/index.php/topic,150639.0.html
dat redesign discussion: http://forum.arcadecontrols.com/index.php/topic,141058.0.html

In any event, no one seems to have done much with controls.dat in eighteen months or so if not longer.

---

At the very least, the previous controls.xml can provide reference for verified button names, which can later be localized.

however, additional features of the format include button colors - also found in yet another file known as colors.ini, for which I cannot recall if we support? - as well as a "Misc. Details" tips/tricks/info field in the XML.

I am not sure if this file has been translated into other languages and would have to check with the former mameplus project and maybe Fujix on that. If they have, it's guaranteed to have been more recently maintained.

AntoPISA of ProgettoSnaps, or his friend motoschifo of ArcadeDatabase, might, *might* have made a more modern unofficial Controls.dat package offline somewhere, where he updated the setnames to current MAME. Otherwise there's foggy memories of hearing about an unofficial edit on MAMEWorld or ArcadeControls, but I may be confusing that with Command.dat.

--------------

Command.dat, aside from including tons of fighting game combos, might also include some relevant data on verified button names from the actual manuals/control panels.

crazyc implemented support of this file through Lua already, but it was a carryover from MEWUI, I believe. There's Japanese and English versions of this DAT file, in both "longhand" and "shorthand" forms

--------------

Plus for those interested in all this, there was a very interesting mention of a "more current" file called "ControlsButtons.ini" by Antropus, author of the utility known as Lightspeed:
http://www.mameworld.info/ubbthreads/showthreaded.php?Cat=&Board=news&Number=355296#Post355296
"I worked some more on my own parser that tries to cross info from MAME, history.dat and catver.ini, hoping that, by deduction, some of the missing controls can be deciphered. Out of the 36000+ games supported in 0.174, only 764 games got no info now. I'm refining the filters as I go but it's a big project because not all info seem to be available to decypher the precise type of control. Some are plain impossible without seeing a picture of the cabinet, because internally the hardware is the same, but in reality they are aesthetically different or used in non-standard ways. "

So in other words he tried to obtain even more button and joystick count/names/etc. by parsing pre-existing files.

I cannot tell if he ever distributed this, or if it was merely generated by his utility.

However, he later gave up on this idea: http://www.mameworld.info/ubbthreads/showthreaded.php?Cat=&Number=355501
"- Eliminated the need of ControlsButtons.ini. Thinking about backwards compatibility with older MAME versions in the future, the controls are now being parsed on the fly, with no significant hit in performance."

Still, starting from his ControlsButtons.ini rather than Controls.xml might be a better jumping off point for a new external file.
